File: src/runtime/storage/creatable-storage-key.ts
/**
 * @license
 * Copyright 2019 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../../platform/assert-web.js';
import {StorageKey} from './storage-key.js';

/**
 * Represents a store that will be created once the recipe is instantiated.
 */
export class CreatableStorageKey extends StorageKey {
  public static readonly protocol = 'create';

  constructor(readonly name: string) {
    super(CreatableStorageKey.protocol);
  }

  toString() {
    return `${CreatableStorageKey.protocol}://${this.name}`;
  }

  childWithComponent(_: string): StorageKey {
    throw new Error('childWithComponent is not available for CreatableStorageKeys');
  }

  subKeyWithComponent(_: string): StorageKey {
    throw new Error('subKeyWithComponent is not available for CreatableStorageKeys');
  }

  static fromString(key: string): CreatableStorageKey {
    const match = key.match(/^create:\/\/([^?]+)$/);
    if (!match) {
      throw new Error(`Not a valid CreatableStorageKey: ${key}.`);
    }
    const [_, name] = match;
    return new CreatableStorageKey(name);
  }
}

File: java/arcs/core/data/CreatableStorageKey.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */
package arcs.core.data

import arcs.core.storage.StorageKey
import arcs.core.storage.StorageKeySpec

/**
 * This class represents a storage key in a compiled [Plan] with 'create' fate.
 */
data class CreatableStorageKey(
  val nameFromManifest: String
) : StorageKey(protocol) {

  override fun toKeyString() = nameFromManifest

  override fun childKeyWithComponent(component: String): StorageKey {
    throw UnsupportedOperationException("CreatableStorageKey is used as a placeholder only.")
  }

  override fun toString(): String = super.toString()

  companion object : StorageKeySpec<CreatableStorageKey> {
    override val protocol = "create"

    private val CREATABLE_STORAGE_KEY_PATTERN =
      ("^([^:^?]*)\$").toRegex()

    override fun parse(rawKeyString: String): CreatableStorageKey {
      val match =
        requireNotNull(CREATABLE_STORAGE_KEY_PATTERN.matchEntire(rawKeyString)) {
          "Not a valid CreatableStorageKey"
        }
      return CreatableStorageKey(match.groupValues[1])
    }
  }
}
